### PR TITLE
feat: Indexed Merkle tree concurrency

### PR DESCRIPTION
**Pull request by vadorovsky**

### CODE REVIEW

File: merkle-tree/concurrent/src/lib.rs [hunk 1]
```diff
@@ -15,19 +15,19 @@ use crate::{
...
+//     pub height: usize,
+//     pub changelog_size: usize,
+//     pub current_changelog_index: usize,
+//     pub roots_size: usize,
+//     pub current_root_index: usize,
+//
+//     pub next_index: usize,
+//     pub sequence_number: usize,
+//     pub rightmost_leaf: [u8; 32],
+// }
```
Comment (ananas-block) [inline merkle-tree/concurrent/src/lib.rs:30]: can this be deleted?

File: merkle-tree/indexed/src/errors.rs [hunk 2]
```diff
@@ -30,14 +35,16 @@ pub enum IndexedMerkleTreeError {
...
-            IndexedMerkleTreeError::IndexHigherThanMax => 3002,
-            IndexedMerkleTreeError::LowElementNotFound => 3003,
-            IndexedMerkleTreeError::LowElementGreaterOrEqualToNewElement => 3004,
-            IndexedMerkleTreeError::NewElementGreaterOrEqualToNextElement => 3005,
+            IndexedMerkleTreeError::IntegerOverflow => 10001,
+            IndexedMerkleTreeError::IndexHigherThanMax => 10002,
+            IndexedMerkleTreeError::LowElementNotFound => 10003,
+            IndexedMerkleTreeError::LowElementGreaterOrEqualToNewElement => 10004,
+            IndexedMerkleTreeError::NewElementGreaterOrEqualToNextElement => 10005,
+            IndexedMerkleTreeError::ChangelogBufferSize(_, _) => 10006,
```
Comment (ananas-block) [inline merkle-tree/indexed/src/errors.rs:43]: is there a convention for error types like this?

Comment (ananas-block) [reply merkle-tree/indexed/src/errors.rs:43]: I know that anchor uses 6000

File: merkle-tree/indexed/src/lib.rs [hunk 3]
```diff
@@ -53,91 +88,14 @@ where
         )?;
         Ok(Self {
             merkle_tree,
+            // current_changelog_index: 0,
+            // changelog_capacity: indexed_change_log_size,
+            // changelog_length: 0,
```
Comment (ananas-block) [inline merkle-tree/indexed/src/lib.rs:93]: leftover?

File: merkle-tree/indexed/src/lib.rs [hunk 4]
```diff
@@ -176,14 +134,64 @@ where
...
         low_leaf_proof: &mut BoundedVec<[u8; 32]>,
     ) -> Result<(), IndexedMerkleTreeError> {
+        let patched_low_element = self.patch_low_element(&low_element)?;
+        // match patched_low_element {
+        //     Some((patched_low_element, patched_low_element_next_value)) => {
+        //         low_element = patched_low_element;
+        //         low_element_next_value = BigUint::from_bytes_be(&patched_low_element_next_value);
+        //     }
+        //     None => {}
+        // }
```
Comment (ananas-block) [inline merkle-tree/indexed/src/lib.rs:190]:
```suggestion
```

File: merkle-tree/indexed/tests/tests.rs [hunk 5]
```diff
@@ -80,9 +80,10 @@ where
     merkle_tree.update(
         usize::from(changelog_index),
         nullifier,
-        low_nullifier,
-        low_nullifier_next_value,
+        low_nullifier.clone(),
+        low_nullifier_next_value.clone(),
         low_nullifier_proof,
+        // &mut CyclicBoundedVec::<RawIndexedElement<usize>>::with_capacity(1),
```
Comment (ananas-block) [inline merkle-tree/indexed/tests/tests.rs:86]: leftover

File: merkle-tree/indexed/src/errors.rs (hunk 2 above)
Comment (vadorovsky) [reply merkle-tree/indexed/src/errors.rs:43]: Not really. I just tried to pick numbers not taken by Anchor

Comment (vadorovsky) [reply merkle-tree/indexed/src/errors.rs:43]: > Not really. I just tried to pick numbers not taken by Anchor

But I've just realized I did a quite bad job here. Not in this PR, 10k is fine. But Anchor uses 100-5000:

https://docs.rs/anchor-lang/latest/anchor_lang/error/enum.ErrorCode.html
https://www.anchor-lang.com/docs/errors

We should make sure we take only ranges starting from 6k or 7k. Let's do it in a separate PR?

File: .github/actions/setup-and-build/action.yml [hunk 6]
```diff
@@ -66,6 +66,7 @@ runs:
         # Check for unstaged changes
         if ! git diff --quiet; then
           echo "There are unstaged changes after build!"
+          git diff
```
Comment (vadorovsky) [inline .github/actions/setup-and-build/action.yml:69]: I want to keep this here - showing `git diff` with unstaged changes helps with debugging.

File: merkle-tree/indexed/src/errors.rs (hunk 2 above)
Comment (ananas-block) [reply merkle-tree/indexed/src/errors.rs:43]: one idea that crossed my mind is to introduce levels:
- hash-set: 7000ers
- Merkle-tree: 8000ers
- Indexed: 9000ers
- account-compression: 10000ers
- compressed-pda: 11000ers
- compressed-token: 12000ers

different idea:
- hash-set: 7000-7100
- Merkle-tree: 7100-7200
- Indexed: 7200-7300
- account-compression: 8000ers
- compressed-pda: 9000ers
- compressed-token: 10000ers

**Review by ananas-block — APPROVED** [review]

**Pull request merged**